# Question 1

We know that the margin vectors are given by:
\begin{eqnarray}
  w \cdot X + b = \pm 1\\
\end{eqnarray}

On replacing $1$ with some constant $\gamma$ we get:
\begin{eqnarray}
  w \cdot X + b = \pm \gamma\\
\end{eqnarray}
On dividing the equation with $\gamma$:
\begin{eqnarray}
  \frac {w}{\gamma} \cdot X + \frac {b}{\gamma} = \pm 1\\
\end{eqnarray}

Since the goal of getting the hyperplane equation is minimising $||w||$, dividing $w$ by a scalar doesn't change anything, the equation of the hyper plane will still end up being $w \cdot X + b = 0$. Hence, changing $1$ to another constant $\gamma$ gives us the same solution.



# Question 2

To prove this equation we'll make use of the following two equations from class:

$$
w = \sum_{j} \alpha_j y_j X_j\\
\sum_{j} \alpha_j y_j = 0
$$

And given,
$$
\rho = \frac {1} {||w||}\\
\frac {1} {\rho} = ||w||\\
\frac {1} {\rho^2} = ||w||^2
$$


The equation of the margin vectors is $w^T x + b = \pm 1$. If $X_i$ is a support vector and $y_i$ is its label, we can write  $w^T X_i + b = y_i$.

On substituting:
\begin{eqnarray}
  w^T X_i + b = y_i\\
  \sum_{j} \alpha_j y_j X_j^T X_i + b = y_i
\end{eqnarray}

Multiplying the equation with $\alpha_i y_i$:
\begin{eqnarray}
  \sum_{j} \alpha_j y_j X_j^T X_i \alpha_i y_i + b \alpha_i y_i = \alpha_i y_i^2
\end{eqnarray}

On summing over $i$:
\begin{eqnarray}
  \sum_{i,j} \alpha_j y_j X_j^T X_i \alpha_i y_i + b \sum_{i}\alpha_i y_i = \sum_{i}\alpha_i y_i^2\\
  \sum_{i,j} \alpha_i\alpha_j y_i y_j X_j^T X_i  + b \sum_{i}\alpha_i y_i = \sum_{i}\alpha_i y_i^2
\end{eqnarray}

Since $y_i$ lies on the margin vector, it equals $\pm 1$ thus $y_i^2 = 1$. Using this and substituting values from the equations derived in class, we get:
\begin{eqnarray}
  \sum_{i,j} \alpha_i\alpha_j y_i y_j X_j^T X_i  + b \sum_{i}\alpha_i y_i &=& \sum_{i}\alpha_i y_i^2\\
  \sum_{i,j} \alpha_i\alpha_j y_i y_j X_j^T X_i  + b \times 0 &=& \sum_{i}\alpha_i\\ 
  \sum_{i,j} \alpha_i\alpha_j y_i y_j X_j^T X_i &=& \sum_{i}\alpha_i\\
  \sum_{j} \alpha_j y_j X_j^T \times \sum_{i} \alpha_i y_i X_i &=& \sum_{i}\alpha_i\\
  w^T  w &=& \sum_{i}\alpha_i\\
  ||w||^2 &=& \sum_{i}\alpha_i\\
  \frac 1 {\rho^2} &=& \sum_{i}\alpha_i
\end{eqnarray}

Hence, proved.



# Question 3

For this question, we will be using Mercer's condition to check the validity of the given functions.

A kernel $K(x,z)$ is said to be valid if it satisfies Mercer's condition:
\begin{eqnarray}
  \iint f(x)K(x,z)f(z)\,dx\,dz \geq 0
\end{eqnarray}

Since, $k_1$ and $k_2$ are valid kernels, we can write:
\begin{eqnarray}
  \iint f(x)k_1(x,z)f(z)\,dx\,dz \geq 0\\
  \iint f(x)k_2(x,z)f(z)\,dx\,dz \geq 0
\end{eqnarray}

A kernel $K(x,z)$ can be written as:
\begin{eqnarray}
K(x,z) &=& \phi(x) \cdot \phi(z)\\
K(x,z) &=& \sum_{i} \phi_i(x)\phi_i(z)
\end{eqnarray}

Thus,
\begin{eqnarray}
k_1(x,z) &=& \sum_{i} \phi_i(x)^{(1)}\phi_i(z)^{(1)}\\
k_2(x,z) &=& \sum_{i} \phi_i(x)^{(2)}\phi_i(z)^{(2)}
\end{eqnarray}

## (a)

\begin{eqnarray}
  k(x,z) = k_1(x,z) + k_2(x,z)
\end{eqnarray}

Substituting this in the Mercer's condition:
\begin{eqnarray}
  \iint f(x)k(x,z)f(z)\,dx\,dz &\geq& 0\\
  \iint f(x)(k_1(x,z) + k_2(x,z))f(z)\,dx\,dz &\geq& 0\\
  \iint f(x)k_1(x,z)f(z)\,dx\,dz + \iint f(x)k_2(x,z)f(z)\,dx\,dz &\geq& 0
\end{eqnarray}

Since both parts of the above sum are greater than zero, the final sum will also be greater that zero, thereby satisfying Mercer's condition. Hence, this function is a valid kernel.

## (b)

\begin{eqnarray}
  k(x,z) = k_1(x,z)k_2(x,z)
\end{eqnarray}

Substituting this and the dot product representations in the Mercer's condition:
\begin{eqnarray}
  \iint f(x)k(x,z)f(z)\,dx\,dz &=& \iint f(x)k_1(x,z)k_2(x,z)f(z)\,dx\,dz\\ 
  &=&\iint f(x)\left(\sum_{i} \phi_i(x)^{(1)}\phi_i(z)^{(1)}\right)\left(\sum_{i} \phi_i(x)^{(2)}\phi_i(z)^{(2)}\right)f(z)\,dx\,dz \\
  &=&\sum_{i} \sum_{i} \iint  \phi_i(x)^{(1)} \phi_i(x)^{(2)} f(x)\,dx \phi_i(z)^{(2)} \phi_i(z)^{(1)} f(z) \,dz \\
  &=& \sum_{i}\sum_{i}\int  \phi_i(x)^{(1)}\phi_i(x)^{(2)}f(x)\,dx\int \phi_i(z)^{(1)}\phi_i(z)^{(2)}f(z)\,dz 
\end{eqnarray}

Since both, the integrals are of the same format, we can replace the variables $x$ and $z$ with another generic variable:
\begin{eqnarray}
  &=&\sum_{i}\sum_{i}\left[\int \phi_i(y)^{(1)}\phi_i(y)^{(2)}f(y)\,dy\right]^2&\geq& 0\\
\end{eqnarray}

A squared value is always greater than zero, and thus it satisfies a condition proving that the given function is a valid kernel.

## (c)

$k(x,z) = h(k_1(x,z))$ where h is a polynomial function with positive co-efficients.

We know that the polynomial function is going to be of the following form:
\begin{eqnarray}
  h(k(x,z)) = c_0 + c_1k(x,z) + c_2k(x,z)^2 + c_3k(x,z)^3 + . . . 
\end{eqnarray}

Now, we can write $k(x,z)^2$ as $k(x,z)k(x,z)$ which is a valid kernel as proved in part (b). Similarly, all higher powers of $k(x,z)$ can be proved to be valid kernels in the same way.

Moreover, a scalar multiplied with a kernel is also a valid kernel function since:
\begin{eqnarray}
  C \times k(x,z) = C \iint f(x)k(x,z)f(z)\,dx\,dz\\
  C\iint f(x)k(x,z)f(z)\,dx\,dz &\geq& 0\\
\end{eqnarray}

Thus, by combining the proofs in (a) and (b), we can declare the function $h(k_1(x,z))$ as a valid kernel.


## (d)

\begin{eqnarray}
  k(x,z) = exp(k(x,z))
\end{eqnarray}

We know that the Taylor expansion of the exponential function is the following:
\begin{eqnarray}
  exp(k(x,z)) = 1 + \frac {k(x,z)}{1!} + \frac {k(x,z)^2}{2!} + \frac {k(x,z)^3}{3!} +  . . . 
\end{eqnarray}

On comparing this with the polynomial structure in part (c), we notice that both the expressions have the same structure, hence the expansion is a polynomial. And as the polynomial function is a valid kernel, so is the exponential function.


## (e)

\begin{eqnarray}
  k(x,z) = exp(\frac {-||x - z||_2^2}{\sigma^2})
\end{eqnarray}

On expanding the numerator (Eucledian distance) we get:
\begin{eqnarray}
  exp(\frac {-||x - z||_2^2}{\sigma^2}) &=& exp(\frac {- ||x||^2 + 2 x \cdot z - ||z||^2 }{\sigma^2})\\
  &=& exp(\frac {-||x||^2}{\sigma^2})exp(\frac {2 x \cdot z}{\sigma^2})exp(\frac {-||z||^2}{\sigma^2})
\end{eqnarray}

The second term is the exponent of a kernel (dot product) and that is also a kernel according to part (d). The first and third terms are functions of $x$ and $z$.

Thus, it can be rewritten as:
\begin{eqnarray}
  exp(\frac {-||x - z||_2^2}{\sigma^2}) &=& exp(\frac {-||x||^2}{\sigma^2})exp(\frac {2 x \cdot z}{\sigma^2})exp(\frac {-||z||^2}{\sigma^2})\\
  &=& g(x)exp(k'(x,z))g(z)
\end{eqnarray}

Now, on substituting the above equation into the Mercer's equation we get:
\begin{eqnarray}
  \iint f(x)k(x,z)f(z)\,dx\,dz &=& \iint f(x)g(x)exp(k'(x,z))g(z)f(z)\,dx\,dz\\ 
\end{eqnarray} 

Writing $f(x)g(x)$ as $F(x)$, $f(z)g(z)$ as $F(z)$ $exp(k'(x,z))$ as $K(x,z)$ and we get:
\begin{eqnarray}
  \iint f(x)g(x)exp(k'(x,z))g(z)f(z)\,dx\,dz &=& \iint F(x)K(x,z)F(z)\,dx\,dz\\ 
\end{eqnarray} 
Which is in the same form as the Mercer equation's original format, thus:
\begin{eqnarray}
 \iint F(x)K(x,z)F(z)\,dx\,dz \geq 0\\ 
\end{eqnarray} 

Thus, the given kernel functions is a valid one.


# Question 4

## (a)

In [24]:
import csv
import numpy as np
from sklearn import svm
import sys

# Enter You Name Here
myname = "Soumi-Chakraborty"

def run_svm():

    # Load data set
    # data_set_path_train = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.train.txt"
    # data_set_path_test = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.test.txt"

    data_set_path_train = "http://www.amlbook.com/data/zip/features.train"
    data_set_path_test = "http://www.amlbook.com/data/zip/features.test"
    
    train_data = np.loadtxt(data_set_path_train)
    ones = np.where(train_data[:, 0] == 1)
    fives = np.where(train_data[:, 0] == 5)
    labels = np.append(ones, fives , axis=1)
    labels = np.sort(labels)
    train_data = (train_data[labels])[0]
    print("Number of training records: %d" % len(train_data))

    # test set
    test_data = np.loadtxt(data_set_path_test)
    ones = np.where(test_data[:, 0] == 1)
    fives = np.where(test_data[:, 0] == 5)
    labels = np.append(ones, fives , axis=1)
    labels = np.sort(labels)
    test_data = (test_data[labels])[0]
    print("Number of testing records: %d" % len(test_data))

    # training the classifier
    X = train_data[:, 1:] # all rows, all columns except the last
    y = train_data[:, 0]  # all rows, only the last column
    svc = svm.SVC(kernel ='linear')
    svc.fit(X, y)

    X_test =  test_data[:, 1:] # all rows, all columns except the last
    y_test = test_data[:,0]  # all rows, only the last column

    # predict results for test set
    predictions = svc.predict(X_test)


    # compute accuracy
    results = predictions == y_test
    accuracy = float(np.count_nonzero(results))/float(len(results))
    print("Accuracy on the test set: %.4f" % accuracy)    

    # number of support vectors
    sv_count = svc.n_support_
    print("Support vectors for each class:", sv_count)
    print("Total number of support vectors:", sv_count.sum())  


if __name__ == "__main__":
    run_svm()

Number of training records: 1561
Number of testing records: 424
Accuracy on the test set: 0.9788
Support vectors for each class: [14 14]
Total number of support vectors: 28


## (b)

In [19]:
import csv
import numpy as np
from sklearn import svm

# Enter You Name Here
myname = "Soumi-Chakraborty"

def run_svm(samples):
  print("Samples:", samples)

  # Load data set
  # data_set_path_train = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.train.txt"
  # data_set_path_test = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.test.txt"

  data_set_path_train = "http://www.amlbook.com/data/zip/features.train"
  data_set_path_test = "http://www.amlbook.com/data/zip/features.test"
  
  
  # choose only those rows with labels == 1 or 5
  # train set
  train_data = np.loadtxt(data_set_path_train)
  ones = np.where(train_data[:, 0] == 1)
  fives = np.where(train_data[:, 0] == 5)
  labels = np.append(ones, fives , axis=1)
  labels = np.sort(labels)
  train_data = (train_data[labels])[0][0:samples, :]
  print("Number of training records: %d" % len(train_data))

  # test set
  test_data = np.loadtxt(data_set_path_test)
  ones = np.where(test_data[:, 0] == 1)
  fives = np.where(test_data[:, 0] == 5)
  labels = np.append(ones, fives , axis=1)
  labels = np.sort(labels)
  test_data = (test_data[labels])[0]
  print("Number of testing records: %d" % len(test_data))

  # training the classifier
  X = train_data[:, 1:] # all rows, all columns except the last
  y = train_data[:, 0]  # all rows, only the last column
  svc = svm.SVC(kernel ='linear')
  svc.fit(X, y)

  X_test =  test_data[:, 1:] # all rows, all columns except the last
  y_test = test_data[:,0]  # all rows, only the last column

  # predict results for test set
  predictions = svc.predict(X_test)

  # compute accuracy
  results = predictions == y_test
  accuracy = float(np.count_nonzero(results))/float(len(results))
  print("Accuracy on the test set: %.4f" % accuracy)    

  # number of support vectors
  sv_count = svc.n_support_
  # print("Support vectors for each class:", sv_count)
  print("Total number of support vectors:", sv_count.sum())  
  print()

if __name__ == "__main__":
  
  sample_count = [50, 100, 200, 800]
  for sample in sample_count:
    run_svm(sample)

Samples: 50
Number of training records: 50
Number of testing records: 424
Accuracy on the test set: 0.9811
Total number of support vectors: 2

Samples: 100
Number of training records: 100
Number of testing records: 424
Accuracy on the test set: 0.9811
Total number of support vectors: 4

Samples: 200
Number of training records: 200
Number of testing records: 424
Accuracy on the test set: 0.9811
Total number of support vectors: 8

Samples: 800
Number of training records: 800
Number of testing records: 424
Accuracy on the test set: 0.9811
Total number of support vectors: 14



## (c)

In [36]:
import csv
import numpy as np
from sklearn import svm

# Enter You Name Here
myname = "Soumi-Chakraborty"

def run_svm(C, Q, errorType='test'):
  print("Q =", Q)

  # Load data set
  # data_set_path_train = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.train.txt"
  # data_set_path_test = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.test.txt"

  data_set_path_train = "http://www.amlbook.com/data/zip/features.train"
  data_set_path_test = "http://www.amlbook.com/data/zip/features.test"
  
  
  # choose only those rows with labels == 1 or 5
  # train set
  train_data = np.loadtxt(data_set_path_train)
  ones = np.where(train_data[:, 0] == 1)
  fives = np.where(train_data[:, 0] == 5)
  labels = np.append(ones, fives , axis=1)
  labels = np.sort(labels)
  train_data = (train_data[labels])[0]

  # test set
  test_data = np.loadtxt(data_set_path_test)
  ones = np.where(test_data[:, 0] == 1)
  fives = np.where(test_data[:, 0] == 5)
  labels = np.append(ones, fives , axis=1)
  labels = np.sort(labels)
  test_data = (test_data[labels])[0]

  # training the classifier
  X = train_data[:, 1:] # all rows, all columns except the last
  y = train_data[:, 0]  # all rows, only the last column
  svc = svm.SVC(kernel='poly', degree=Q, coef0=1, gamma=1, C=C)
  svc.fit(X, y)

  X_test =  test_data[:, 1:] # all rows, all columns except the last
  y_test = test_data[:,0]  # all rows, only the last column

  # predict results for test set and training set
  if errorType == 'Test':
    predictions = svc.predict(X_test)
    y_check = y_test
  elif errorType == 'Train':
    predictions = svc.predict(X)
    y_check = y

  # error
  results = predictions == y_check
  accuracy = float(np.count_nonzero(results))/float(len(results))
  error = 1 - accuracy
  print(str(errorType), "error: %.4f" % error)    

  # number of support vectors
  sv_count = svc.n_support_
  print("Support vectors:", sv_count.sum())  
  print()


if __name__ == "__main__":
  
  print("Part i.")
  run_svm(C=0.0001, Q=2, errorType='Train')
  run_svm(C=0.0001, Q=5, errorType='Train')

  print("Part ii.")
  run_svm(C=0.001, Q=2, errorType='Train')
  run_svm(C=0.001, Q=5, errorType='Train')

  print("Part iii.")
  run_svm(C=0.01, Q=2, errorType='Train')
  run_svm(C=0.01, Q=5, errorType='Train')

  print("Part iv.")
  run_svm(C=1, Q=2, errorType='Test')
  run_svm(C=1, Q=5, errorType='Test')

Part i.
Q = 2
Train error: 0.0090
Support vectors: 236

Q = 5
Train error: 0.0045
Support vectors: 26

Part ii.
Q = 2
Train error: 0.0045
Support vectors: 76

Q = 5
Train error: 0.0045
Support vectors: 25

Part iii.
Q = 2
Train error: 0.0045
Support vectors: 34

Q = 5
Train error: 0.0038
Support vectors: 23

Part iv.
Q = 2
Test error: 0.0189
Support vectors: 24

Q = 5
Test error: 0.0212
Support vectors: 21



i. False\
ii. True\
iii. False\
iv. False



## (d)

In [22]:
import csv
import numpy as np
from sklearn import svm

# Enter You Name Here
myname = "Soumi-Chakraborty"

def run_svm(C):
  print("C =", C)

  # Load data set
  # data_set_path_train = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.train.txt"
  # data_set_path_test = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/features.test.txt"

  data_set_path_train = "http://www.amlbook.com/data/zip/features.train"
  data_set_path_test = "http://www.amlbook.com/data/zip/features.test"
  
  
  # choose only those rows with labels == 1 or 5
  # train set
  train_data = np.loadtxt(data_set_path_train)
  ones = np.where(train_data[:, 0] == 1)
  fives = np.where(train_data[:, 0] == 5)
  labels = np.append(ones, fives , axis=1)
  labels = np.sort(labels)
  train_data = (train_data[labels])[0]

  # test set
  test_data = np.loadtxt(data_set_path_test)
  ones = np.where(test_data[:, 0] == 1)
  fives = np.where(test_data[:, 0] == 5)
  labels = np.append(ones, fives , axis=1)
  labels = np.sort(labels)
  test_data = (test_data[labels])[0]

  # training the classifier
  X = train_data[:, 1:] # all rows, all columns except the last
  y = train_data[:, 0]  # all rows, only the last column
  svc = svm.SVC(kernel='rbf', gamma=1, C=C)
  svc.fit(X, y)

  X_test =  test_data[:, 1:] # all rows, all columns except the last
  y_test = test_data[:,0]  # all rows, only the last column

  # predict results for test set and training set
  predictions_test = svc.predict(X_test)
  predictions_train = svc.predict(X)

  # test error
  results = predictions_test == y_test
  accuracy = float(np.count_nonzero(results))/float(len(results))
  error = 1 - accuracy
  print("Test error: %.4f" % error)

  # train error
  results = predictions_train == y
  accuracy = float(np.count_nonzero(results))/float(len(results))
  error = 1 - accuracy
  print("Train error: %.4f" % error)

  print()


if __name__ == "__main__":
  
  cVal = [0.01, 1, 100, 10000, 1000000]
  for C in cVal:
    run_svm(C)

C = 0.01
Test error: 0.0236
Train error: 0.0038

C = 1
Test error: 0.0212
Train error: 0.0045

C = 100
Test error: 0.0189
Train error: 0.0032

C = 10000
Test error: 0.0236
Train error: 0.0026

C = 1000000
Test error: 0.0236
Train error: 0.0006



Lowest test error: 1.89%\
Given by C = 100

Lowest train error: 0.06%\
Given by C = 1000000

# Question 5

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


## (a)

In [4]:
import csv
import numpy as np
from sklearn import svm
import sys

# Enter You Name Here
myname = "Soumi-Chakraborty"

def run_svm():

    # Load data set
    train_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_train.data"
    train_labels_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_train.labels"
    valid_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_valid.data"
    valid_labels_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_valid.labels"

    # train_path = "gisette_train.data"
    # train_labels_path = "gisette_train.labels"
    # valid_path = "gisette_valid.data"
    # valid_labels_path = "gisette_valid.labels"

    # train set
    X = np.loadtxt(train_path)
    y = np.loadtxt(train_labels_path)
    print("Number of training records: %d" % len(X))

    # test set
    X_test = np.loadtxt(valid_path)
    y_test = np.loadtxt(valid_labels_path)
    print("Number of test records: %d" % len(X_test))
    
    # training the classifier
    svc = svm.SVC(kernel ='linear')
    svc.fit(X, y)

    # predict results for test set and training set
    predictions_test = svc.predict(X_test)
    predictions_train = svc.predict(X)

    # test error
    results = predictions_test == y_test
    accuracy = float(np.count_nonzero(results))/float(len(results))
    error = 1 - accuracy
    print("Test error: %.4f" % error)
    

    # train error
    results = predictions_train == y
    accuracy = float(np.count_nonzero(results))/float(len(results))
    error = 1 - accuracy
    print("Train error: %.4f" % error)
    
    # number of support vectors
    sv_count = svc.n_support_
    print("Total number of support vectors:", sv_count.sum())

    


if __name__ == "__main__":
    run_svm()

Number of training records: 6000
Number of test records: 1000
Test error: 0.0240
Train error: 0.0000
Total number of support vectors: 1084


## (b)

In [5]:
import csv
import numpy as np
from sklearn import svm
import sys

# Enter You Name Here
myname = "Soumi-Chakraborty"


def run_svm(kernel):
  print("Kernel type:", kernel)


  # Load data set
  train_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_train.data"
  train_labels_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_train.labels"
  valid_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_valid.data"
  valid_labels_path = "/content/drive/MyDrive/Acad_Stuff/Course Work/Semester 5/FoML/Ass2/gisette_valid.labels"

  # train_path = "gisette_train.data"
  # train_labels_path = "gisette_train.labels"
  # valid_path = "gisette_valid.data"
  # valid_labels_path = "gisette_valid.labels"
  
  # train set
  X = np.loadtxt(train_path)
  y = np.loadtxt(train_labels_path)
  print("Number of training records: %d" % len(X))

  # test set
  X_test = np.loadtxt(valid_path)
  y_test = np.loadtxt(valid_labels_path)
  print("Number of test records: %d" % len(X_test))
  
  # training the classifier
  if kernel == 'rbf':
    svc = svm.SVC(kernel ='rbf', gamma=0.001)
    svc.fit(X, y)
  elif kernel == 'poly':
    # svc = svm.SVC(kernel ='poly', gamma=1, degree=2, coef0=1)
    svc = svm.SVC(kernel ='poly', degree=2, coef0=1)
    svc.fit(X, y)
  else:
    svc = svm.SVC(kernel ='linear')
    svc.fit(X, y)

  # predict results for test set and training set
  predictions_test = svc.predict(X_test)
  predictions_train = svc.predict(X)

  # test error
  results = predictions_test == y_test
  accuracy = float(np.count_nonzero(results))/float(len(results))
  error = 1 - accuracy
  print("Test error: %.4f" % error)


  # train error
  results = predictions_train == y
  accuracy = float(np.count_nonzero(results))/float(len(results))
  error = 1 - accuracy
  print("Train error: %.4f" % error)

  # number of support vectors
  sv_count = svc.n_support_
  print("Total number of support vectors:", sv_count.sum())
  print()

    


if __name__ == "__main__":
    run_svm('rbf')
    run_svm('poly')


Kernel type: rbf
Number of training records: 6000
Number of test records: 1000
Test error: 0.5000
Train error: 0.0000
Total number of support vectors: 6000

Kernel type: poly
Number of training records: 6000
Number of test records: 1000
Test error: 0.0200
Train error: 0.0005
Total number of support vectors: 1332



The rbf kernel reports the lowest training error since it's 0 whereas the polynomial kernel reports a training error of 0.0005.